### PR TITLE
[inductor] generate triton kernel benchmark

### PR DESCRIPTION
**Pull request by shunting314**
A PR to generate benchmark code for individual triton kernels. We can explore improving autotuning with the saved compiled kernel directly. This potentially can speedup our iteration and separate the concern with the upstream components that generate the compiled module.

Since I'm still ramping up on inductor, I'll reflect what I learned here so people can correct me if I'm wrong.  In inductor, WrapperCodeGen class is used to generate the compiled module for CUDA (or triton). Here is an example compiled module for a toy model like: `def f(x): return sin(x) + cos(x)` https://gist.github.com/shunting314/c6ed9f571919e3b414166f1696dcc61b .  A compiled module contains the following part:
- various triton kernels
- a wrapper (or a method named call . The name is hardcoded) that calls the triton kernels and potentially ATen kernels to efficiently do the same work as the original Fx graph being compiled by inductor
- some utility code that generate random inputs and run the wrapper

The triton kernels in the compiled module are annotated with decorator like pointwise which is used for autotuning.

This PR add a config so enabling it will just trigger the path of the compiled module being printed. It can be controlled from environment variable as well.

The path to each compiled triton kernel is added as comment in the compiled module. E.g.
```
# kernel path: /tmp/torchinductor_shunting/gn/cgn6x3mqoltu7q77gjnu2elwfupinsvcovqwibc6fhsoiy34tvga.py
triton__0 = async_compile.triton('''
import triton
import triton.language as tl
...
""")
````

Example command:
```
TORCHINDUCTOR_OUTPUT_COMPILED_MODULE_PATH=1 TORCHINDUCTOR_BENCHMARK_KERNEL=1 python benchmarks/dynamo/huggingface.py --backend inductor --amp --performance --training --dashboard --only AlbertForMaskedLM --disable-cudagraphs
```

cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Comment by shunting314** [comment]
Change plans.

File: torch/_inductor/codegen/triton.py [hunk 1]
```diff
@@ -1171,6 +1172,49 @@ def codegen_body(self):
...
+            result.writeline("from torch._inductor.triton_ops.autotune import grid")
+            result.writeline("")
+
+            argdefs, call_args, signature = self.args.python_argdefs()
+
+            for arg_name in call_args:
+                buf = V.graph.get_buffer(arg_name)
+                if buf:
+                    result.writeline(
+                        f"{arg_name} = rand_strided({tuple(buf.get_size())}, {tuple(buf.get_stride())}, device='{buf.get_device()}', dtype={buf.get_dtype()})"
```
Comment (ngimel) [inline torch/_inductor/codegen/triton.py:1194]: this won't be exactly right for aliased inputs/outputs, but I don't think it has material impact on perf

File: torch/_inductor/codegen/triton.py [hunk 2]
```diff
@@ -1171,6 +1172,49 @@ def codegen_body(self):
...
+
+            call_args = ", ".join(call_args)
+            index = V.graph.scheduler.current_device.index
+            stream_name = f"stream{index}"
+            result.writeline(f"{stream_name} = get_cuda_stream({index})")
+
+            # print_performance assert non-None return value, but a triton kernel
+            # indeed returns None. Override the return value to True to handle this.
+            result.writeline(
+                f"print_performance(lambda: triton_.run({call_args}, grid=grid({', '.join(grid)}), stream={stream_name}) or True)"
```
Comment (ngimel) [inline torch/_inductor/codegen/triton.py:1212]: @chillee I'm not sure we want `print_performance` here, should `TORCHINDUCTOR_PROFILE` be enough?

Comment (shunting314) [reply torch/_inductor/codegen/triton.py:1212]: AFAICT, `TORCHINDUCTOR_PROFILE` collects latency and bandwidth information for each triton kernel being called. With that, we can find good candidates (with pure perf) to dive into.

For the kernel benchmark in this PR, I can let the benchmark print latency and bandwidth information as well. This way, we can easily tune configs manually and see the impact on the latency/bandwidth of the kernel.

Comment (Chillee) [reply torch/_inductor/codegen/triton.py:1212]: imo we should ideally be able to have some flexibility in how we call the kernel and with how many arguments.

For example, ideally, the code would look something like
```
def call(args):
    arg0_1, = args
    args.clear()
    with torch.cuda._DeviceGuard(0):
        torch.cuda.set_device(0) # no-op to ensure context
        buf0 = empty_strided((2, 3), (3, 1), device='cuda', dtype=torch.float32)
        stream0 = get_cuda_stream(0)
        triton__0.run(arg0_1, buf0, 6, grid=grid(6), stream=stream0)
        del arg0_1
        return (buf0, )

def get_args():
      arg0_1 = rand_strided((2, 3), (3, 1), device='cuda:0', dtype=torch.float32)
      return (arg0_1,)

if __name__ == "__main__":
    from torch._dynamo.testing import rand_strided
    from torch._inductor.utils import print_performance
    args = get_args()
    print_performance(lambda: call(list(args)))
```

And then future work would be able to do something like
```
for kernel in kernel_dataset:
    # do something with kernel
```

File: torch/_inductor/graph.py [hunk 3]
```diff
@@ -590,8 +597,8 @@ def compile_to_module(self):
         for name, value in self.constants.items():
             setattr(mod, name, value)
 
-        if dynamo_config.output_code:
-            log.info("Output code: %s", mod.__file__)
+        if config.output_compiled_module_path:
```
Comment (Chillee) [inline torch/_inductor/graph.py:600]: I'm also a little bit uncertain about what's changing here and why we need 2 environmental variables for this?

File: torch/_inductor/codegen/triton.py (hunk 2 above)
Comment (shunting314) [reply torch/_inductor/codegen/triton.py:1212]: So given a kernel, its argument list is already fixed. Currently the generated benchmark code looks like:

```
if __name__ == "__main__":
    from torch._C import _cuda_getCurrentRawStream as get_cuda_stream
    from torch._dynamo.testing import rand_strided
    from torch._inductor.utils import print_performance
    import torch
    from torch._inductor.triton_ops.autotune import grid

    arg0_1 = rand_strided((2, 3), (3, 1), device='cuda:0', dtype=torch.float32)
    buf0 = rand_strided((2, 3), (3, 1), device='cuda:0', dtype=torch.float32)
    stream0 = get_cuda_stream(0)
    print_performance(lambda: triton_.run(arg0_1, buf0, 6, grid=grid(6), stream=stream0) or True)
    ```

I can add the missing DeviceGuard. One more difference I see in your example is, you treat arg0_1 as kernel input and buf0 as kernel output (although it's address is passed in as kernel argument). But in my example, they are alls randomly generated arguments. I think your way is better. The question I have is, do we have a good way to figure out which kernel arguments is actually for output?

File: torch/_inductor/graph.py (hunk 3 above)
Comment (shunting314) [reply torch/_inductor/graph.py:600]: So if we enable output_code config other things like
- guards
- fx graph
are also printed.

Added another environment variable so when enabling it, we just specifically print the compiled module path.

File: torch/_inductor/codegen/triton.py (hunk 2 above)
Comment (Chillee) [reply torch/_inductor/codegen/triton.py:1212]: @shunting314 I'm saying that I think the main thing that's missing from this PR is the ability to pull out the kernel from the dumped file and have control over how often I'm calling it.

Like, if you import the kernel file,  you'll have access to a call function, but we don't have a way of calling that function since we don't have programmatic access to the arguments. Our only way of calling the function is to call `__main__`.

Comment (shunting314) [reply torch/_inductor/codegen/triton.py:1212]: ah got it.  I'll organized code into
- call function
- get_args function
- use the above 2 functions in `__main__`

File: torch/_inductor/graph.py (hunk 3 above)
Comment (Chillee) [reply torch/_inductor/graph.py:600]: I guess it seems weird to me to add 2 new env variables for one piece of new functionality.  In particular, this change seems weird since we're just changing some logging info for the pre-existing config option.

Comment (shunting314) [reply torch/_inductor/graph.py:600]: I see. Would it be better if I only add config.benchmark_kernel . Print the compiled module path if that config is true?

BTW, the reason I change log.info to print is because enable info log just cause too much other stuff being being printed. So I print directly to stderr.

**Comment by shunting314** [comment]
Some test on vgg16 shows this PR does not handle random seed input properly. I'll need debug more. Otherwise it's ready for reviewing again.

Re Horace's another comment about doing sth like
```
for kernel in kernel_dataset:
    # do something with kernel
```
I think we can add an option to the `__main__` of the compiled module so it can either
- run the compiled module's call function (the current behavior)
- or run the benchmark for all the kernels in this file.

I can add that in a followup PR if people don't mind :)

File: torch/_inductor/codegen/triton.py [hunk 4]
```diff
@@ -1171,6 +1172,70 @@ def codegen_body(self):
...
+            )
+            result.writeline("from torch._dynamo.testing import rand_strided")
+            result.writeline("from torch._inductor.utils import get_num_bytes")
+            result.writeline("import torch")
+            result.writeline("from torch._inductor.triton_ops.autotune import grid")
+            result.writeline("from triton.testing import do_bench")
+            result.writeline("")
+
+            result.writeline(
+                "ms = do_bench(lambda: call(get_args()), rep=40, fast_flush=True)[0]"
```
Comment (Chillee) [inline torch/_inductor/codegen/triton.py:1228]: I think you should move the `get_args()` out of the lambda. As it is, you'll benchmark the argument construction code every iteration as well.

**Review by Chillee — APPROVED** [review]
Overall LGTM. Looks quite useful, and a good start :)

File: torch/_inductor/codegen/triton.py (hunk 4 above)
Comment (shunting314) [reply torch/_inductor/codegen/triton.py:1228]: Good catch, will do!

**Comment by shunting314** [comment]
The random seed issue mentioned above is fixed in latest revision of this PR. The reason is random seed is treated as constant tensor rather than buffer in GraphLowering class. Have to handle that separately when generating kernel benchmarks.

**Review by Chillee — APPROVED** [review]
